prover/newprover.go

```diff
@@ -0,0 +1,37 @@
+package prover
+
+import (
+	"github.com/spacemeshos/merkle-tree"
+	"github.com/spacemeshos/merkle-tree/cache"
+	"github.com/spacemeshos/poet/shared"
+)
+
+func GetProof(challenge shared.NewChallenge, leafCount uint64, securityParam uint8) (shared.MerkleProof, error) {
+	treeCache := cache.NewWriter(
+		cache.Combine(
+			cache.SpecificLayersPolicy(map[uint]bool{0: true}),
+			cache.MinHeightPolicy(11)),
+		cache.MakeSliceReadWriterFactory())
+	tree := merkle.NewTreeBuilder().WithHashFunc(challenge.MerkleHashFunc()).WithCacheWriter(treeCache).Build()
```

Comment:
The hash function is derived from the challenge (the challenge is used as salt). The verifier should use the expected challenge (the root if the commitment tree) and see if the root is identical.

If you're wondering why I'm using an interface for the challenge rather than pass the real object, it's because the existing code supports different hash functions (sha256 / scrypt) and I know this is not finalized yet, so I wanted to support easily benchmarking different functions.

Now that I'm thinking about it some more - I may as well hard code it and benchmark a different branch for each hash function, since it's not something we'll want to make configurable eventually.